internal/evmreader/claim.go

```diff
@@ -0,0 +1,224 @@
+				previousEpochs, err := r.repository.GetPreviousSubmittedClaimsEpochs(
+					ctx, app, claimAcceptance.LastProcessedBlockNumber.Uint64())
+				if err != nil {
+					slog.Error("Error retrieving previous submitted claims",
+						"app", app, "error", err)
+					continue APP_LOOP
+				}
+
+				for _, previousEpoch := range previousEpochs {
+					previousEpoch.Status = EpochStatusClaimRejected
+					epochs = append(epochs, previousEpoch)
+					slog.Warn("Claim rejected",
+						"app", app,
+						"lastBlock", previousEpoch.LastBlock,
+						"hash", previousEpoch.ClaimHash)
+				}
+
+				// Get the Epoch for the current Claim Acceptance Event
+				epoch, err := r.repository.GetEpoch(
+					ctx, calculateEpochIndex(
+						r.epochLengthCache[app],
+						claimAcceptance.LastProcessedBlockNumber.Uint64()),
+					app)
+				if err != nil {
+					slog.Error("Error retrieving Epoch", "app", app, "error", err)
+					continue APP_LOOP
+				}
+
+				// Check Epoch
+				if epoch == nil {
+					slog.Error("Got claim acceptance event for an unknown epoch",
+						"app", app,
+						"claim last block", claimAcceptance.LastProcessedBlockNumber,
+						"hash", claimAcceptance.Claim)
+					continue APP_LOOP
+				}
+
+				// Update Epoch claim status
+				if claimAcceptance.Claim != *epoch.ClaimHash ||
+					claimAcceptance.LastProcessedBlockNumber.Uint64() != epoch.LastBlock {
+					slog.Warn("Claim Rejected",
+						"app", app,
+						"lastBlock", epoch.LastBlock,
+						"hash", epoch.ClaimHash)
+
+					epoch.Status = EpochStatusClaimRejected
+					epochs = append(epochs, epoch)
+				} else {
+					slog.Info("Claim Accepted",
+						"app", app,
+						"lastBlock", epoch.LastBlock,
+						"hash", epoch.ClaimHash)
+
+					epoch.Status = EpochStatusClaimAccepted
+					epochs = append(epochs, epoch)
+				}
+			}
+
+			// Store everything
+			err = r.repository.UpdateEpochsClaimStatusTransaction(
+				ctx, app, epochs, mostRecentBlockNumber)
+			if err != nil {
+				slog.Error("Error storing claims", "app", app, "error", err)
+				continue
+			}
+
+		}
+
+	}
+
+}
+
+func (r *EvmReader) readClaimAcceptanceFromBlockchain(
+	ctx context.Context,
+	consensusContract ConsensusContract,
+	appAddresses []common.Address,
+	startBlock, endBlock uint64,
+) (map[common.Address][]*iconsensus.IConsensusClaimAcceptance, error) {
+	appClaimAcceptanceMap := make(map[common.Address][]*iconsensus.IConsensusClaimAcceptance)
+	for _, address := range appAddresses {
+		appClaimAcceptanceMap[address] = []*iconsensus.IConsensusClaimAcceptance{}
+	}
+	opts := &bind.FilterOpts{
+		Context: ctx,
+		Start:   startBlock,
+		End:     &endBlock,
+	}
+	claimAcceptanceEvents, err := consensusContract.RetrieveClaimAcceptanceEvents(
+		opts, appAddresses)
+	if err != nil {
+		return nil, err
+	}
+	for _, event := range claimAcceptanceEvents {
+		appClaimAcceptanceMap[event.AppContract] = insertSorted(
+			sortByLastBlockNumber, appClaimAcceptanceMap[event.AppContract], event)
+	}
```

Comment:
You can remove the extra loop by checking the map contents before inserting, with the added benefit of removing from the map addresses of apps with no events.

```suggestion
	opts := &bind.FilterOpts{
		Context: ctx,
		Start:   startBlock,
		End:     &endBlock,
	}
	claimAcceptanceEvents, err := consensusContract.RetrieveClaimAcceptanceEvents(
		opts, appAddresses)
	if err != nil {
		return nil, err
	}
	for _, event := range claimAcceptanceEvents {
		if _, ok := appClaimAcceptanceMap[event.AppContract]; ok {
			appClaimAcceptanceMap[event.AppContract] = insertSorted(
				sortByLastBlockNumber,
				appClaimAcceptanceMap[event.AppContract],
				event,
			)
		} else {
			appClaimAcceptanceMap[event.AppContract] = insertSorted(
				sortByLastBlockNumber,
				[]*iconsensus.IConsensusClaimAcceptance{event},
				event,
			)
		}
	}
```

Reply:
Even applications with no Acceptance Events have it's `LastClaimCheckBlock`updated at each run. That's why they are in the map